lib/MySQL_Monitor.cpp

```diff
@@ -0,0 +1,768 @@
+#include <map>
+#include <list>
+#include <thread>
+#include "proxysql.h"
+#include "cpp.h"
+
+
+#ifdef DEBUG
+#define DEB "_DEBUG"
+#else
+#define DEB ""
+#endif /* DEBUG */
+#define MYSQL_MONITOR_VERSION "0.2.0519" DEB
+
+
+#include <event2/event.h>
+
+extern ProxySQL_Admin *GloAdmin;
+extern MySQL_Threads_Handler *GloMTH;
+
+
+static MySQL_Monitor *GloMyMon;
+
+#define NEXT_IMMEDIATE(new_st) do { ST= new_st; goto again; } while (0)
+
+static void state_machine_handler(int fd, short event, void *arg);
+
+
+/*
+struct state_data {
+	int ST;
+	char *hostname;
+	int port;
+	struct event ev_mysql;
+	MYSQL mysql;
+	MYSQL_RES *result;
+	MYSQL *ret;
+	int err;
+	MYSQL_ROW row;
+	struct query_entry *query_element;
+	int index;
+};
+*/
+
+static int connect__num_active_connections;
+static int total_connect__num_active_connections=0;
+static int ping__num_active_connections;
+static int total_ping__num_active_connections=0;
+
+
+struct cmp_str {
+	bool operator()(char const *a, char const *b)
+	{
+		return strcmp(a, b) < 0;
+	}
+};
+
+class MySQL_Monitor_Connection_Pool {
+	private:
+	int size;
+	//std::map<std::pair<char *, std::list<MYSQL *>* > my_connections;
+	std::map<char *, std::list<MYSQL *>* , cmp_str> my_connections;
+	public:
+	MySQL_Monitor_Connection_Pool();
+	~MySQL_Monitor_Connection_Pool();
+	MYSQL * get_connection(char *hostname, int port);
+	void put_connection(char *hostname, int port, MYSQL *my);
+};
+
+MySQL_Monitor_Connection_Pool::MySQL_Monitor_Connection_Pool() {
+	size=0;
+}
+
+MySQL_Monitor_Connection_Pool::~MySQL_Monitor_Connection_Pool() {
+}
+
+MYSQL * MySQL_Monitor_Connection_Pool::get_connection(char *hostname, int port) {
+	std::map<char *, std::list<MYSQL *>* >::iterator it;
+	//it = my_connections.find(std::make_pair(hostname,port));
+	char *buf=(char *)malloc(16+strlen(hostname));
+	sprintf(buf,"%s:%d",hostname,port);
+	it = my_connections.find(buf);
+	free(buf);
+	if (it != my_connections.end()) {
+		std::list<MYSQL *> *lst=it->second;
+		if (!lst->empty()) {
+			MYSQL *ret=lst->front();
+			lst->pop_front();
+			size--;
+			return ret;
+		}
+	}
+	return NULL;
+}
+
+void MySQL_Monitor_Connection_Pool::put_connection(char *hostname, int port, MYSQL *my) {
+	size++;
+	std::map<char *, std::list<MYSQL *>* >::iterator it;
+	//std::map<std::pair<char *, int>, std::list<MYSQL *>* >::iterator it;
+	char * buf=(char *)malloc(16+strlen(hostname));
+	sprintf(buf,"%s:%d",hostname,port);
+	it = my_connections.find(buf);
+	//it = my_connections.find(std::make_pair(hostname,port));
+	std::list<MYSQL *> *lst=NULL;
+	if (it==my_connections.end()) {
+		lst=new std::list<MYSQL *>;
+		//my_connections.insert(std::pair<std::pair<char *,int>(hostname,port), std::list<MYSQL *>*>(lst));
+		my_connections.insert(my_connections.begin(), std::pair<char *,std::list<MYSQL *>*>(buf,lst));
```

Comment:
I think if you're planning to roll out more STL usage, using std::string will make the code more readable going forward. It should be quite fast (and there are also drop-in replacements -- e.g. https://news.ycombinator.com/item?id=8704684).